### ISSUE TITLE
fs : read file sync after sync copy of directory

### ISSUE THREAD

**Issue by giper45** [nodejs/help#622]
* **Version**: v7.8.0
* **Platform**: MacOS Sierra

Hello, could you explain what's happening with this example code? It is a recursive copy of dirs, than 2 reads in files that exists inside the copied dir. 
 It seems that the recursive copy is finished inside the before function (by looking at the console.log('dirs created'), and then  the it functions  starts. It doesn't gives no file error, it gives empty buffer result, but when I open the copied dir I find 2 files correctly written. I wrote this simple example because this happened in strange situations inside my async code (with no builtin recursive copy, but by using some npm modules that does it) and I tried to reproduce in a simple example this issue. The code is this: 
```javascript

const chai = require('chai');
const path = require('path');
const fs = require('fs');
const jsonfile = require('jsonfile');
chai.expect;
const mkdir = (dir) => {
  // making directory without exception if exists
  try {
    fs.mkdirSync(dir);
  } catch (e) {
    if (e.code !== 'EEXIST') {
      throw e;
    }
  }
};
const rmdir = (dir) => {
  if (fs.existsSync(dir)) {
    const list = fs.readdirSync(dir);
    for (let i = 0; i < list.length; i += 1) {
      const filename = path.join(dir, list[i]);
      const stat = fs.statSync(filename);

      if (filename === '.' || filename === '..') {
      // pass these files
      } else if (stat.isDirectory()) {
      // rmdir recursively
        rmdir(filename);
      } else {
      // rm fiilename
        fs.unlinkSync(filename);
      }
    }
    fs.rmdirSync(dir);
  } else {
    console.warn(`warn: ${dir} not exists`);
  }
};

const copy = (src, dest) => {
  const oldFile = fs.createReadStream(src);
  const newFile = fs.createWriteStream(dest);
  oldFile.pipe(newFile);
};
const copyDir = (src, dest) => {
  mkdir(dest);
  const files = fs.readdirSync(src);
  for (let i = 0; i < files.length; i += 1) {
    const current = fs.lstatSync(path.join(src, files[i]));
    if (current.isDirectory()) {
      copyDir(path.join(src, files[i]), path.join(dest, files[i]));
    } else if (current.isSymbolicLink()) {
      const symlink = fs.readlinkSync(path.join(src, files[i]));
      fs.symlinkSync(symlink, path.join(dest, files[i]));
    } else {
      copy(path.join(src, files[i]), path.join(dest, files[i]));
    }
  }
};

let otherData;
describe('Test', function() {
  before(function()  {
    rmdir('test_project');
    copyDir('/oldPath/test_project', 'test_project');
    console.log('dir created');
    this.data = fs.readFileSync('test_project/lab_states.json');
    otherData = fs.readFileSync('test_project/example.txt');
  });
  it('Sync test', function()  {
    console.log(this.data);
    console.log(otherData);
  });
  it('Sync test', function()  {
    console.log(this.data);
    console.log(otherData);
  });
  it('Sync test', function()  {
    console.log(this.data);
    console.log(otherData);
  });
  it('Sync test', function()  {
    console.log(this.data);
    console.log(otherData);
  });
... other it functions
``` 
The result is: 
```
  Test
dir created
<Buffer >
<Buffer >
    ✓ Sync test
<Buffer >
<Buffer >
    ✓ Sync test
<Buffer >
<Buffer >
    ✓ Sync test
<Buffer >
<Buffer >
    ✓ Sync test
<Buffer >
<Buffer >
    ✓ Sync test
<Buffer >
<Buffer >
    ✓ Sync test
<Buffer >
<Buffer >
    ✓ Sync test
<Buffer >
<Buffer >
    ✓ Sync test
<Buffer >
<Buffer >
    ✓ Sync test
<Buffer >
<Buffer >
    ✓ Sync test
```  
Any advices? Thanks in advance.

**Comment by gireeshpunathil** [comment]
@giper45 - before reviewing the code and make any inferences, just want to know if this is something still outstanding?

**Comment by giper45** [comment]
Hello @gireeshpunathil, if you look at https://github.com/jprichardson/node-jsonfile/issues/79 the issue is still open and someone still post the issue ;
 I think that this could be an fs module problem, even if this issue doesn't seem to have the same 'popularity' as the Jsonfile issue...

**Comment by gireeshpunathil** [comment]
thanks, I will have a look and let you know

**Comment by gireeshpunathil** [comment]
@giper45 - I reviewed the code. Basically most of your program lines are `sync` methods so everything should execute in sequence. 

the only exception to this is your copy function - though it is invoked from between `sync` methods,  the usage of streams make the copy operation `async`.

So when you actually perform the readSync method to get the file data, there was nothing in it, and the empty buffer made sense, but by the time you manually inspect the files, the stream operations was concluded, and hence you see proper data.

workaround is to make the data copy also synchronous:

```patch
40,42c40
<   const oldFile = fs.createReadStream(src);
<   const newFile = fs.createWriteStream(dest);
<   oldFile.pipe(newFile);
---
>   fs.copyFileSync(src, dest);
```

I have verified this to be working as expected. Other workaround of wrapping the file read operation within a setTimeout call is feasible, bug prone to race conditions.

Hope this helps.
